pkg/controllers/node/node_deleter.go

```diff
@@ -0,0 +1,103 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package node
+
+import (
+	"context"
+	"time"
+
+	client "github.com/projectcalico/libcalico-go/lib/clientv3"
+	cerrors "github.com/projectcalico/libcalico-go/lib/errors"
+	"github.com/projectcalico/libcalico-go/lib/options"
+	log "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/util/workqueue"
+)
+
+// NewNodeDeletionController creates a new controller responsible for garbage collection Calico node objects
+// in etcd mode when their corresponding Kubernetes node is deleted.
+func NewNodeDeletionController(client client.Interface, cs *kubernetes.Clientset) *nodeDeleter {
+	return &nodeDeleter{
+		clientset: cs,
+		client:    client,
+		rl:        workqueue.DefaultControllerRateLimiter(),
+	}
+}
+
+type nodeDeleter struct {
+	rl        workqueue.RateLimiter
+	clientset *kubernetes.Clientset
+	client    client.Interface
+}
+
+func (c *nodeDeleter) RegisterWith(f *DataFeed) {
+	// No-op - we only care about Kubernetes node deletion events.
+}
+
+func (c *nodeDeleter) OnKubernetesNodeDeleted() {
+	// When a Kubernetes node is deleted, trigger a sync.
+	log.Debug("Kubernetes node deletion event")
+	c.deleteStaleNodes()
+}
+
+// deleteStaleNodes compares the set of Calico and Kubernetes nodes and deletes and nodes
+// which are present in the Calico datastore, but whose corresponding Kubernetes node has
+// been deleted. Non-Kubernetes nodes will be ignored.
+func (c *nodeDeleter) deleteStaleNodes() error {
+	// Possibly rate limit calls to Calico
+	time.Sleep(c.rl.When(RateLimitCalicoList))
+	cNodes, err := c.client.Nodes().List(context.TODO(), options.ListOptions{})
+	if err != nil {
+		log.WithError(err).Error("Error listing Calico nodes")
+		return err
+	}
+	c.rl.Forget(RateLimitCalicoList)
+
+	time.Sleep(c.rl.When(RateLimitK8s))
+	kNodes, err := c.clientset.CoreV1().Nodes().List(context.Background(), metav1.ListOptions{})
+	if err != nil {
+		log.WithError(err).Error("Error listing K8s nodes")
+		return err
+	}
+	c.rl.Forget(RateLimitK8s)
+	kNodeIdx := make(map[string]bool)
+	for _, node := range kNodes.Items {
+		kNodeIdx[node.Name] = true
+	}
+
+	for _, node := range cNodes.Items {
+		k8sNodeName, err := getK8sNodeName(node)
+		if err != nil {
+			if _, ok := err.(*ErrorNotKubernetes); ok {
+				log.WithError(err).Info("Skipping non-kubernetes node")
+				continue
+			}
+			log.WithError(err).Error("Error getting k8s node name")
+			continue
+		}
+		if k8sNodeName != "" && !kNodeIdx[k8sNodeName] {
+			// No matching Kubernetes node with that name
+			time.Sleep(c.rl.When(RateLimitCalicoDelete))
```

Comment:
Hm, I think calico/node will only re-create the node instance if it is restarted, so the node instance would be lost in this case. I think this is probably it's own bug that we need to fix - it will have other implications as well. For example, deleting and re-creating a Node in KDD mode will lose most of the metadata we store on the node object (BGP IP at least, from my testing). 

It's still probably worth making this code do a re-query of the node immediately before deleting so that we close that window from potentially seconds down to milliseconds. But that's not really a fix - calico/node needs to be spot when a node has gone down and fix the missing data or else we end up with a broken node.